src/core/HibpDownloader.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ *  Copyright (C) 2020 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "HibpDownloader.h"
+#include "core/Config.h"
+#include "core/Global.h"
+#include "core/NetworkManager.h"
+
+#include <QCryptographicHash>
+#include <QUrl>
+#include <QtNetwork>
+
+namespace
+{
+
+    /*
+     * Return the SHA1 hash of the specified password in upper-case hex.
+     *
+     * The result is always exactly 40 characters long.
+     */
+    QString sha1Hex(const QString& password)
+    {
+        // Get the binary SHA1
+        const auto sha1 = QCryptographicHash::hash(password.toUtf8(), QCryptographicHash::Sha1);
+
+        // Convert to hex
+        QString ret;
+        ret.reserve(40);
+        auto p = sha1.data();
+        for (int i = 0; i < sha1.size(); ++i) {
+            const auto u = uint8_t(*p++);
+            constexpr auto hex = "0123456789ABCDEF";
+            ret += hex[(u >> 4) & 0xF];
+            ret += hex[u & 0xF];
+        }
+
+        return ret;
+    }
+
+    /*
+     * Search a password's hash in the output of the HIBP web service.
+     *
+     * Returns the number of times the password is found in breaches, or
+     * 0 if the password is not in the HIBP result.
+     */
+    int pwnCount(const QString& password, const QString& hibpResult)
+    {
+        const auto hash = sha1Hex(password).toStdString();
+        const auto result = hibpResult.toStdString();
+        const auto pHash = hash.c_str();
+        const auto pResult = result.c_str();
+
+        // The first 5 characters of the hash are in the URL. Search the
+        // rest in the HIBP result.
+        const auto p = strstr(pResult, pHash + 5);
+        if (p) {
+
+            // Found: Return the number after the next colon
+            const auto colon = strchr(p, ':');
+            return colon ? atoi(colon + 1) : 1;
+
+        } else {
+
+            // Not found
+            return 0;
+        }
+    }
+} // namespace
+
+HibpDownloader::HibpDownloader(QString password, QObject* parent)
```

Comment:
Instead of just passing in 1 password, you should pass in a QStringList of passwords and shove them all into the network manager. It will batch process them and fire off a signal when each is completed. You combine each reply with the password that made it to do the post processing.

Look at the custom icon batch downloader code for guidance.
https://github.com/keepassxreboot/keepassxc/blob/b045160e4fcd612def4c5cff55f9469baf12a738/src/gui/IconDownloaderDialog.cpp#L62-L103